[Enterprise Search] Internal Connector representation

jedrazb opened the pull request:
<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

This PR implements the internal representation of a `Connector` doc. Information from multiple places is combined for the final representation of the object, relevant resources:
- [kbn-search-connectors/types/connectors.ts](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/connectors.ts#L216)
- [CONNECTOR_PROTOCOL.md](https://github.com/elastic/connectors/blob/main/docs/CONNECTOR_PROTOCOL.md)
- [elastic-connector-mappings.json index template](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-mappings.json)

## Changes
- Add internal representation of a connector doc: `Connector` class. (except for `configuration` object, that's left out for next PR)
- Add single, basic endpoint: `PUT` that allows for creation of connector with given ID 
- Added `connector_id` field to the connector doc index template, we need to copy the `_id` value inside the doc to enable for e.g. sorting of connector that is needed for `list` endpoint
- Unit tests added
- Manually tested

jedrazb committed:
[ent-search] Start implementation of Connectors API, define the Connector class and work on a simple PUT api

jedrazb committed:
DELETE _connector/<id> method implemented

jedrazb committed:
Add more fields to Connector class

jedrazb committed:
Add ConnectorScheduling class with toXContent function

jedrazb committed:
Add more instance variables, use Builder pattern, move enums to new classes

jedrazb committed:
Connector representation almost done

jedrazb committed:
Add some tests

jedrazb committed:
Merge branch 'main' into jedrazb-connectors-api-connectos-object-and-basic-endpoints

jedrazb commented on x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-mappings.json:
```diff
@@ -18,6 +18,9 @@
         "api_key_id": {
           "type": "keyword"
         },
+        "connector_id": {
```
It felt useful to add for e.g. sorting the paged response of `list` endpoint. I had issues getting `_id` to work for sorting.

kderusso replied:
To my knowledge copying the `_id` field to a regular document field is still the [best practice](https://discuss.elastic.co/t/sort-by-id-field/169017) 👍

carlosdelest replied:
I was able to sort via ID on synonyms [here](https://github.com/elastic/elasticsearch/blob/874fbbc2f05b201873edc21b5baa36753935296f/server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java#L214).

sphilipse replied:
Great call, we've been doing this in Kibana internally for convenience and it makes everything a lot easier. Doing this in the API makes a lot of sense.

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java:
```diff
@@ -0,0 +1,176 @@
+    );
+
+    static {
+        PARSER.declareBoolean(constructorArg(), FILTERING_ADVANCED_CONFIG_ENABLED_FIELD);
```
This parser works fine processing the JSON body from and `updateFeatures` endpoint (that is yet to be implemented, but I tested it with put endpoint). The [elastic-connectors index template](https://github.com/elastic/elasticsearch/blob/b03939ec9d3fb90ac83ab62dcc3219cfbe41e536/x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-mappings.json#L33) forces us to write thr data to index as a complex, nested structure, see `toXContent` method in this class. 

I think the problematic case would be reading the features stored in an index and representing them as a `ConnectorFeatures` instance. In that case, we would need another parser that can work with deserialising XContent data stored in an index - but given its nested nature it's not straightforward to define for me. 

Am I correct to assume that this will only be an issue if we want to support partial updates to the individual fields?

carlosdelest replied:
I'm not sure I understand correctly. Do you mean to have an external representation (API) that is different than the internal representation (index)?

I would advise to create separate classes and parsing methods for that. We do something similar in SearchApplications; storage is done [here](https://github.com/elastic/elasticsearch/blob/b7eafce32cb1e7dfdcb259aab89b886ff93a1c66/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java#L283) and read is done [here](https://github.com/elastic/elasticsearch/blob/b7eafce32cb1e7dfdcb259aab89b886ff93a1c66/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java#L428). It can be quite confusing as we do some binary parsing there, but the main point is we can use different classes for de/serialization

jedrazb replied:
> Do you mean to have an external representation (API) that is different than the internal representation (index)?

Yes, that's what I meant. And thank you for linking the examples :)

kderusso commented on x-pack/plugin/ent-search/build.gradle:
```diff
@@ -14,7 +14,8 @@ base {
 dependencies {
-  compileOnly project(path: xpackModule('core'))
+  implementation project(path: ':libs:elasticsearch-x-content')
+    compileOnly project(path: xpackModule('core'))
```
Nitpick: Indentation

jedrazb replied:
Not sure why it was changed, fixed :)

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -98,6 +107,8 @@ public class EnterpriseSearch extends Plugin implements ActionPlugin, SystemInde
 
     public static final String QUERY_RULES_API_ENDPOINT = "_query_rules";
 
+    public static final String CONNECTORS_API_ENDPOINT = "_connector";
```
Naming nitpick 

```suggestion
    public static final String CONNECTOR_API_ENDPOINT = "_connector";
```

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -140,6 +151,12 @@ protected XPackLicenseState getLicenseState() {
             new ActionHandler<>(PutQueryRulesetAction.INSTANCE, TransportPutQueryRulesetAction.class),
+
+            // Connectors
+            new ActionHandler<>(DeleteConnectorAction.INSTANCE, TransportDeleteConnectorAction.class),
```
No GET?

jedrazb replied:
The logic of GET was slightly more complex than the list endpoint, so I left it out for future PRs. Carlos suggested that I just focus on adding a single, simple PUT endpoint as a part of this PR, and I will follow this approach.

kderusso replied:
That's a good suggestion and the pattern we followed when adding APIs for search relevance 👍

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchBaseRestHandler.java:
```diff
@@ -28,6 +28,9 @@ protected EnterpriseSearchBaseRestHandler(XPackLicenseState licenseState, Licens
     protected final BaseRestHandler.RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
         if (LicenseUtils.supportedLicense(this.licenseState)) {
             return innerPrepareRequest(request, client);
+        } else if (this.product.equals(LicenseUtils.Product.CONNECTOR)) {
```
Rather than adding this check, why not have the connectors handlers just implement `BaseRestHandler`?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+    public Connector(StreamInput in) throws IOException {
+        this.connectorId = in.readString();
+        this.apiKeyId = in.readOptionalString();
+        if (in.readBoolean()) {
```
Why not call [readOptionalMap](https://github.com/elastic/elasticsearch/blob/9063088350388e5318be852730eb1560c1682f7a/server/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java#L397) here?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+        try (XContentParser parser = XContentHelper.createParser(XContentParserConfiguration.EMPTY, source, xContentType)) {
+            return Connector.fromXContent(connectorId, parser);
+        } catch (IOException e) {
+            throw new ElasticsearchParseException("Failed to parse: " + source.utf8ToString(), e);
```
Are there PII concerns logging this source? I see things like API keys so we might want to be careful here.

jedrazb replied:
That's correct, good spot! The configuration might contain client secrets (api keys to external sources). I will limit the amount of information we log.

jedrazb replied:
I realised that I don't need parser for the `Connector` class - as it deals only with internal representation of the data in the index. 

I will keep your comment in mind in subsequent PRs that deal with passing/updating PIIs :+1:

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+ *     <li>...</li>
+ * </ul>
+ */
+public class Connector implements Writeable, ToXContentObject {
```
I think we're going to need to bump transport version here. It's probably good to start with the version we're using and do checks to ensure that we're on or after the appropriate version.

carlosdelest replied:
Clusters will either support this API or not for now, it will be all or nothing. I don't think we'll need to bump transport version unless there are changes in the actual class serialization itself. That will only happen if this serialization evolves over time.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorCustomSchedule.java:
```diff
@@ -0,0 +1,161 @@
+    private final String name;
+
+    private ConnectorCustomSchedule(
+        Map<String, Object> configurationOverrides,
```
What's the data we anticipate storing in overrides? Is it structured enough to warrant something more granular than `Map<String,Object>`?

jedrazb replied:
This `configurationOverrides` can override anything related to crawler settings. The fact that crawler is mixed up with Connector doc is a result of an old project that aimed to represent crawler as a connector (it's not done yet). `configurationOverrides ` can be a pretty complex object and I would prefer to avoid going down the rabbit hole to represent this correctly here. What do you think, is this approach acceptable?

jedrazb replied:
Actually, I took a deeper look into documentation and example docs and this object seems feasible to implement: https://www.elastic.co/guide/en/enterprise-search/8.9/crawler-managing.html#crawler-managing-multiple-schedules will add its representation :+1:

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+    @Nullable
+    private final String apiKeyId;
+    @Nullable
+    private final Map<String, Object> configuration;
```
What is configuration composed of? Is `Map<String,Object>` necessary here or should this be more structured?

carlosdelest replied:
++ for adding explicit types!

jedrazb replied:
Is it acceptable to defer defining explicit types until the next PR? Regarding the 'update configuration' endpoint, there is a proposal to separate the configuration schema from its values. I believe it would simplify the review process if I tackled this in a dedicated PR alongside the update configuration endpoint changes.

carlosdelest replied:
I think that's fair @jedrazb . Just make a note / TODO so we understand there's follow up work on this 💯

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:
```diff
@@ -0,0 +1,137 @@
+        }
+    }
+
+    public void deleteConnector(String connectorId, ActionListener<DeleteResponse> listener) {
```
We have a pretty nice pattern of creating a `DelegatingIndexNotFoundActionListener` - this allows streamlining of the listeners in functions like this and delegating failures. Here's an [example](https://github.com/elastic/elasticsearch/blob/9063088350388e5318be852730eb1560c1682f7a/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java#L480).

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIngestPipeline.java:
```diff
@@ -0,0 +1,142 @@
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBoolean(extractBinaryContent);
+        out.writeOptionalString(name);
```
I think you want `writeString` here

jedrazb replied:
Good catch, all args are required here, confirmed it.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/RestDeleteConnectorAction.java:
```diff
@@ -0,0 +1,44 @@
+
+import static org.elasticsearch.rest.RestRequest.Method.DELETE;
+
+public class RestDeleteConnectorAction extends EnterpriseSearchBaseRestHandler {
```
Here are the Action classes that you'll want to change to extend `BaseRestHandler` so you don't have to worry about license checks

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:
```diff
@@ -18,7 +18,8 @@ public final class LicenseUtils {
         SEARCH_APPLICATION("search application"),
         BEHAVIORAL_ANALYTICS("behavioral analytics"),
-        QUERY_RULES("query rules");
+        QUERY_RULES("query rules"),
+        CONNECTOR("connector");
```
```suggestion
        QUERY_RULES("query rules");
```

kderusso reviewed (commented):
Looks great so far! I completed an initial passthrough. You'll want to make sure to add BWC serialization tests for all of your requests/responses as well.

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+            this.configuration = null;
+        }
+        if (in.readBoolean()) {
+            this.customScheduling = new ConnectorCustomSchedule(in);
```
readOptionalWriteable would work here as well

jedrazb replied:
Thank you for pointing this out 👍 I missed that's it's possible :)

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+        this.lastAccessControlSyncError = in.readOptionalString();
+        this.lastAccessControlSyncScheduledAt = in.readOptionalString();
+        if (in.readBoolean()) {
+            this.lastAccessControlSyncStatus = ConnectorSyncStatus.valueOf(in.readString());
```
readOptionalWriteable would work here as well. Applicable to other classes down below 👍

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+        this.indexName = indexName;
+        this.isNative = isNative;
+        this.language = language;
+        this.lastAccessControlSyncError = lastAccessControlSyncError;
```
I wonder, would it make sense to group some of these fields into classes? We can serialize them as a non-nested JSON object, but it would maybe simplify code and serialization management. I can think of creating a SyncStatus class that would contain:

        String lastAccessControlSyncError,
        String lastAccessControlSyncScheduledAt,
        Long lastDeletedDocumentCount,
      String lastSyncError,
        String lastSyncScheduledAt,
        ConnectorSyncStatus lastSyncStatus,
        String lastSynced,
        Object syncCursor,
        boolean syncNow
        String lastSyncError,
        String lastSyncScheduledAt,
        Long lastIndexedDocumentCount,

jedrazb replied:
Yeah I was actually thinking about the same when I was looking at the `Connector` class. I didn't realise that we can serialise this as as non-nested structure. 👍  That will simplify things a bit :)

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+    private static final ConstructingObjectParser<Connector, String> PARSER = new ConstructingObjectParser<>(
+        "connector",
+        false,
+        (args, connectorId) -> new Builder().setConnectorId(connectorId)
```
As mentioned, it might make sense to break this class into additional subclasses for easier parsing and attribute management.

jedrazb replied:
Agreed that looks ugly ...

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+        out.writeOptionalString(description);
+        out.writeOptionalString(error);
+
+        out.writeBoolean(features != null);
```
Pair with readOptionalWriteable; use writeOptionalWriteable here

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,753 @@
+            return this;
+        }
+
+        public Connector createConnector() {
```
```suggestion
        public Connector build() {
```

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:
```diff
@@ -0,0 +1,137 @@
+
+    public static final String CONNECTOR_INDEX_NAME = ConnectorTemplateRegistry.CONNECTOR_INDEX_NAME_PATTERN;
+
+    public ConnectorIndexService(Client clientWithOrigin, ClusterSettings clusterSettings) {
+        this.clientWithOrigin = clientWithOrigin;
```
I would rather receive a regular `Client` and create the `OriginSettingClient` in the constructor, like [here](https://github.com/elastic/elasticsearch/blob/b7eafce32cb1e7dfdcb259aab89b886ff93a1c66/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java#L113):

```suggestion
    public ConnectorIndexService(Client clientWithOrigin, ClusterSettings clusterSettings) {
        this.clientWithOrigin = new OriginSettingClient(client, CONNECTORS_ORIGIN);
```

jedrazb replied:
The connectors-index is not a system index. In that case, should we even care about creating `clientWithOrigin`?

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIngestPipeline.java:
```diff
@@ -0,0 +1,142 @@
+    private final Boolean reduceWhitespace;
+    private final Boolean runMlInference;
+
+    private ConnectorIngestPipeline(Boolean extractBinaryContent, String name, Boolean reduceWhitespace, Boolean runMlInference) {
```
Are these params nullable? Be sure to check that in the constructor if they are not

jedrazb replied:
Do you mean to add explicit assertion checks in the constuctor in case some field is not nullable?

carlosdelest replied:
Yes 👍 . Fail fast should be preferred in this cases

jedrazb replied:
Added `Objects.requireNotNull` checks

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIngestPipeline.java:
```diff
@@ -0,0 +1,142 @@
+
+    public static class Builder {
+
+        private Boolean extractBinaryContent;
```
Are there default values for these booleans? Make sure that the builder (and class constructor) respects that

jedrazb replied:
Do you mean to add explicit assertion checks in the constuctor?

jedrazb replied:
No default values here according to how [we do it in Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/lib/create_connector_document.ts#L45). It's callers responsibility to set it up with appropriate update call

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorScheduling.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    public ConnectorScheduling(StreamInput in) throws IOException {
+        this.accessControl = readScheduleConfigFromStream(in);
```
You can use the StreamInput constructors for the classes to read directly from the StreamInput:

```suggestion
        this.accessControl = new ScheduleConfig(in);
```

as the constructors consume from the StreamInput, it's equivalent to the methods you've defined here 👍

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorScheduling.java:
```diff
@@ -0,0 +1,203 @@
+            return this;
+        }
+
+        public ConnectorScheduling createConnectorScheduling() {
```
```suggestion
        public ConnectorScheduling build() {
```

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIngestPipeline.java:
```diff
@@ -0,0 +1,142 @@
+            return this;
+        }
+
+        public ConnectorIngestPipeline createConnectorIngestPipeline() {
```
```suggestion
        public ConnectorIngestPipeline build() {
```

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java:
```diff
@@ -0,0 +1,176 @@
+            return this;
+        }
+
+        public ConnectorFeatures createConnectorFeatures() {
```
```suggestion
        public ConnectorFeatures build() {
```

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorCustomSchedule.java:
```diff
@@ -0,0 +1,161 @@
+            return this;
+        }
+
+        public ConnectorCustomSchedule createConnectorCustomSchedule() {
```
```suggestion
        public ConnectorCustomSchedule build() {
```

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorScheduling.java:
```diff
@@ -0,0 +1,203 @@
+                return this;
+            }
+
+            public ScheduleConfig createScheduleConfig() {
```
```suggestion
            public ScheduleConfig build() {
```

carlosdelest reviewed (commented):
Great job, @jedrazb !

First pass review. It would help to separate this PR into a single action (PUT), add all tests needed to it, and then focus on individual PRs for the other actions. That would make reviews and changes easier.

Keep the good work!

jedrazb commented:
Thank you for first reviews! I highly appreciate your feedback! 

> It would help to separate this PR into a single action (PUT), add all tests needed to it, and then focus on individual PRs for the other actions. That would make reviews and changes easier.

That's actually a great suggestion, I will remove other methods in this PR so that we can solely focus on correct representation of `Connector` internally.

jedrazb committed:
[WIP] Address review feedback, delete some API method to limit the PR scope

Mikep86 commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorCustomSchedule.java:
```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    public static class ConfigurationOverride implements Writeable, ToXContentObject {
```
Is this a placeholder definition for `ConfigurationOverride`? It looks to be specific to the crawler connector.

jedrazb replied:
This is a definition (not a placeholder IMO) of an allowed configuration override. This is only used by crawler (and not connectors). In the past there was some effort to unify representation of a generic connector and crawler in a single connector doc - hence it's a bit mixed up right now as we are in an in-between state. 

Here is the link to details: https://www.elastic.co/guide/en/enterprise-search/8.9/crawler-managing.html#crawler-managing-multiple-schedules

Mikep86 replied:
Got it, thanks for the background!

jedrazb committed:
Keep working on review feedback, add tests

jedrazb committed:
Add more tests, fill missing equals() and hashCode() methods

Mikep86 commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,473 @@
+
+    private final Map<String, Object> configuration;
+
+    private final ConnectorCustomSchedule customScheduling;
```
If I understand the [custom scheduling docs](https://www.elastic.co/guide/en/enterprise-search/8.9/crawler-managing.html#crawler-managing-multiple-schedules-creating-multiple-schedules) correctly, shouldn't this be a `Map<String, ConnectorCustomSchedule>`, so that you can have multiple custom schedules?

jedrazb replied:
Yes! That's a good spot 👍 I realised it while adding unit tests and just pushed a fix. Thank you for spotting bugs ;)

Mikep86 commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorCustomSchedule.java:
```diff
@@ -0,0 +1,292 @@
+
+public class ConnectorCustomSchedule implements Writeable, ToXContentObject {
+
+    private final Map<String, ConfigurationOverride> configurationOverrides;
```
If I understand the [custom scheduling docs](https://www.elastic.co/guide/en/enterprise-search/8.9/crawler-managing.html#crawler-managing-multiple-schedules-creating-multiple-schedules) correctly, shouldn't this be just `ConfigurationOverride`? There is only one configuration override associated with a custom schedule.

jedrazb replied:
Yes! That's a good spot 👍 I realised it while adding unit tests and just pushed a fix.

jedrazb committed:
Add javadocs, add more tests, fix bug with configuration overrides

jedrazb committed:
Merge branch 'main' into jedrazb-connectors-api-connectos-object-and-basic-endpoints

jedrazb committed:
Fix failing unit test

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:Enterprise Search)

jedrazb committed:
Fix bug introduced in the last commit

timgrein commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:
```diff
@@ -0,0 +1,50 @@
+        try {
+            final IndexRequest indexRequest = new IndexRequest(CONNECTOR_INDEX_NAME).opType(DocWriteRequest.OpType.INDEX)
+                .id(connector.getConnectorId())
+                .opType(DocWriteRequest.OpType.INDEX)
```
Just spotted it, while copying things line by line 👀  `opType` is set twice here.

jedrazb replied:
Thank you for spotting, the credit goes to https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/QueryRulesIndexService.java#L237 from where I copied this 😃

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java:
```diff
@@ -0,0 +1,229 @@
+        builder.startObject();
+        {
+            if (documentLevelSecurityEnabled != null) {
+                builder.startObject("document_level_security");
```
we can reuse DOCUMENT_LEVEL_SECURITY_ENABLED_FIELD.getPreferredName() here to be consistent

jedrazb replied:
I re-used names where I could, but the nested nature of some older feature fields (lovely tech debt ;) ) makes it required to use different names. 

This parser will be use for parsing json body coming from an API call in future PRs.

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:
```diff
@@ -0,0 +1,50 @@
+    public static final String CONNECTOR_INDEX_NAME = ConnectorTemplateRegistry.CONNECTOR_INDEX_NAME_PATTERN;
+
+    public ConnectorIndexService(Client client) {
+        this.client = client;
```
Do we need to use both clients internally? We should use just the clientWithOrigin, right?

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java:
```diff
@@ -0,0 +1,261 @@
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(LAST_ACCESS_CONTROL_SYNC_ERROR.getPreferredName(), lastAccessControlSyncError);
```
Same thing about writing only non-null elements

carlosdelest commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfoTests.java:
```diff
@@ -0,0 +1,50 @@
+import static java.util.Collections.emptyList;
+import static org.hamcrest.CoreMatchers.equalTo;
+
+public class ConnectorSyncInfoTests extends ESTestCase {
```
I see you've done `AbstractBWCSerializationTestCase`s later on. Great work! 👏 

Are these individual tests used for the classes that are used for the internal index representation and not included into request / responses? If so, you did double great work 👏 👏

jedrazb replied:
The internal classes for now are only used only for internal index representation and are not included in request/response. In the next PRs:
- `Connector` would be returned from `GET` request (to be implemented)
- Some classes, that were used to represent connector properties, will overlap with our desired API schema (request) - in that case I will just implement parsers there and will use them for parsing request JSON. In that case I will add thorough parsing <-> toXContnet tests to see if representation works fine :)

carlosdelest replied:
Thanks for clarifying! 👍

carlosdelest reviewed (commented):
This is amazing work Jedr. You've got a huge model here, and you have diligently worked through a ton of boilerplate to make it done. 🔝 

Besides some non-blocking thoughts, I think two things would be necessary for this PR to be merged:
- Add this API behind a feature flag
- Include YAML tests (see [this example](https://github.com/elastic/elasticsearch/blob/c9a2555f09b4c9148b8422cbcaf3f81416eb92cc/x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml#L1-L0)) for having an end-to-end test of the APIs. We can do this in a separate PR if you prefer.

Almost there! Keep the good work!

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,443 @@
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        {
+            builder.field(ID_FIELD.getPreferredName(), connectorId);
```
Ahh I got slightly confused by the fact that in our status quo we were setting some fields explicitly to `null` ([code reference](https://github.com/elastic/kibana/blob/b335c3abbcd776a5d9d6b4e042e2165478cc42ae/packages/kbn-search-connectors/lib/create_connector_document.ts#L45)) ... but I agree that we can avoid writing nulls to index

sphilipse replied:
Yeah that's just a Kibana TypeScript convention, we don't need to actually write those here :)

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorListItem.java:
```diff
@@ -0,0 +1,58 @@
+    private final String name;
+
+    public ConnectorListItem(String connectorId, @Nullable String name) {
+        this.connectorId = connectorId;
```
Good spot! Thank you

jedrazb commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorCustomSchedule.java:
```diff
@@ -0,0 +1,331 @@
+    private final Boolean enabled;
+    private final String interval;
+    @Nullable
+    private final String lastSynced;
```
<img width="548" alt="Screenshot 2023-11-22 at 16 02 17" src="https://github.com/elastic/elasticsearch/assets/14121688/4bfc7c8b-40a9-4bb5-b570-e29d352e5d38">

According to [Elastic docs](https://www.elastic.co/guide/en/enterprise-search/8.9/crawler-managing.html#crawler-managing-multiple-schedules) it's a String (?) 

But I see that for [Filtering we should be handling data as Dates ](https://github.com/elastic/elasticsearch/blob/05b2591d0342d9735fd5f88bffe741ffd83a9607/x-pack/plugin/core/template-resources/src/main/resources/entsearch/connector/elastic-connectors-mappings.json#L74C20-L76C23) I will fix it there

jedrazb replied:
I took a look at [CONNECTOR_PROTOCOL.md](https://github.com/elastic/connectors/blob/main/docs/CONNECTOR_PROTOCOL.md) search for "advanced snippets", fields like `created_at` or `updated_at` are expected to be a string

```
                "advanced_snippet": {
                  "created_at": "2023-11-09T15:13:08.231Z",
                  "updated_at": "2023-11-09T15:13:08.231Z",
                  "value": {}
                },
```
I think it can stay as a `String`

sphilipse replied:
Ah, it's a string there because Elasticsearch automatically parses date strings. It makes more sense to have it represented as a date internally.

carlosdelest replied:
but it's not a String, right? It's a timestamp. Don't we want to parse that to ensure it's correctly formatted, and create the actual timestamp on the server side?

sphilipse replied:
Yes exactly @carlosdelest. The difference is that the connector protocol was an external represenation, and this is an internal representation, so these should be dates.

jedrazb replied:
It seems that `Instant` class is widely used in the codebase to handle time/dates, I will use this to represent time related data

jedrazb commented:
@carlosdelest Thank you for quickly reviewing the code! Your help is invaluable 👍 One question about the naming convention, the tests are prefixed with a number, I read the [docs](https://github.com/elastic/elasticsearch/blob/main/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/README.asciidoc) and [general testing docs](https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#testing-the-rest-layer) but I can't find a good resource explaining what does the prefix number mean exactly and what should I set it for my yaml test?

carlosdelest commented:
>  the tests are prefixed with a number, I read the [docs](https://github.com/elastic/elasticsearch/blob/main/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/README.asciidoc) and [general testing docs](https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#testing-the-rest-layer) but I can't find a good resource explaining what does the prefix number mean exactly and what should I set it for my yaml test?

The number is BASIC style, just for having some ordering in the list of tests (not running order but logical order). I suggest you start using `10_` and go up from there when adding more tests 👍

sphilipse commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,443 @@
+            && Objects.equals(pipeline, connector.pipeline)
+            && Objects.equals(scheduling, connector.scheduling)
+            && Objects.equals(serviceType, connector.serviceType)
+            && status == connector.status
```
question: just curious, why is this a different equality check? It's not the only string in here so I would guess that's not it.

jedrazb replied:
This is autogenerated by Intellij. The status is represented internally as enum. I asked chatgpt about it, here is the response (I'm not even pretending that I'm such a Java expert 😅  )

> When comparing enum values in Java, it's best to use the == operator rather than Object.equals. This is because enums in Java are inherently singletons, meaning that there's only one instance of each enum constant. Due to this singleton nature, using == is safe and more performant for comparing enum values, as it compares the object references directly.

sphilipse reviewed (approved):
Great job! Exciting to see this continue. I hope your wrists are okay with all this typing.

jedrazb committed:
Hide the API behind feature flag, add JavaDocs, use booleans where possible, don't write nulls to index, change to private where possible

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -120,29 +127,42 @@ protected XPackLicenseState getLicenseState() {
+            )
         );
+
+        // Check if Connector API is enabled
```
```suggestion
        // Connectors
```

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -120,29 +127,42 @@ protected XPackLicenseState getLicenseState() {
+        // Check if Connector API is enabled
+        if (ConnectorAPIFeature.isEnabled()) {
+            actionHandlers.add(
+                // Connectors
```
```suggestion
```

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -160,27 +180,39 @@ public List<RestHandler> getRestHandlers(
+            )
         );
+
+        // Check if Connector API is enabled
```
```suggestion
        // Connectors
```

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:
```diff
@@ -160,27 +180,39 @@ public List<RestHandler> getRestHandlers(
+        // Check if Connector API is enabled
+        if (ConnectorAPIFeature.isEnabled()) {
+            restHandlers.add(
+                // Connectors
```
```suggestion
```

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,477 @@
+        this.pipeline = pipeline;
+        this.scheduling = scheduling;
+        this.serviceType = serviceType;
+        this.status = status;
```
Do we want to validate `status` is non-null here, or define a default? (`status` is not annotated as `@Nullable`)

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,477 @@
+    @Nullable
+    private final String apiKeyId;
+    @Nullable
+    private final Map<String, Object> configuration; // TODO: add explicit types
```
Do you plan to add explicit types in a followup PR?

jedrazb replied:
Yes this will be added in a followup PR

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,477 @@
+            if (syncCursor != null) {
+                builder.field(SYNC_CURSOR_FIELD.getPreferredName(), syncCursor);
+            }
+            if (status != null) {
```
`status` is currently not annotated as Nullable

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:
```diff
@@ -0,0 +1,477 @@
+    @Nullable
+    private final Object syncCursor;
+    @Nullable
+    private final boolean syncNow;
```
I don't think you should use a `@Nullable` annotation here. If you truly need to support null use `Boolean`.

jedrazb replied:
That's right, I forgot to remove it after I changed `Boolean` to `boolean`

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorCustomSchedule.java:
```diff
@@ -0,0 +1,333 @@
+            private Integer maxCrawlDepth;
+            private Boolean sitemapDiscoveryDisabled;
+            private List<String> domainAllowList;
+            private List<String> sitemapUrls;
```
Question: Can Sitemap URLs be duplicated? Any use case for using a `Set` here instead of a `List` so we don't double our work?

jedrazb replied:
Using `Set` might be justified to deal with potential duplicates. Given that this data is used in Kibana to render user-provided input I think preserving ordering is important for UX, so I would prefer to keep it as a `List`

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java:
```diff
@@ -0,0 +1,239 @@
+public class ConnectorFeatures implements Writeable, ToXContentObject {
+
+    @Nullable
+    private final Boolean documentLevelSecurityEnabled;
```
Would it be worth setting these to their default values instead of null if not specified? (Disclaimer: I have no strong feelings about this question at all).

jedrazb replied:
From my understanding the values are optional and tied to definitions in connectors framework. Can we leave it as it is, I will revisit this when implementing the endpoint to update those features?

sphilipse replied:
I think setting these booleans to false is perfectly fine. Making them nullable just adds complication without adding much value.

They're nullable in the protocol because these properties were added at later stages, so we had to take account documents that didn't have this property at all. But we have more control in this API so that's not really a concern, and we can just set them to `false` instead.

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFiltering.java:
```diff
@@ -0,0 +1,591 @@
+
+        private final String advancedSnippetCreatedAt;
+        private final String advancedSnippetUpdatedAt;
+        private final Map<String, Object> advancedSnippetValue;
```
Do we intend to create a more granular class for this `Map<String,Object>`?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFiltering.java:
```diff
@@ -0,0 +1,591 @@
+        }
+    }
+
+    public enum FilteringRuleRule {
```
Nitpick: Can we come up with a name that doesn't have duplication? Maybe something like `FilteringRuleCondition`?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorListItem.java:
```diff
@@ -0,0 +1,60 @@
+
+    private final String connectorId;
+
+    private final String name;
```
I think we want to annotate this with `@Nullable`?

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/PutConnectorAction.java:
```diff
@@ -0,0 +1,253 @@
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            {
+                builder.field("description", description);
```
Question, why are we not encoding the connector id into xcontent here?

jedrazb replied:
The `AbstractBWCSerializationTestCase` requires the `PutConnectorAction.Request` to implement the toXContent method. I didn't include the `connectorId` as it's not parsed directly from JSON body. 

It seems that toXContent is only used for BWC tests, it's not used in the codebase for anything else

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/TransportPutConnectorAction.java:
```diff
@@ -0,0 +1,71 @@
+            .setServiceType(request.getServiceType())
+
+            // Default values related to Connector creation process
+            .setStatus(isNative ? ConnectorStatus.NEEDS_CONFIGURATION : ConnectorStatus.CREATED)
```
I would like to see setting these defaults when creating the Connector itself, rather than sending in the request

jedrazb replied:
Moved this to `Connector.Builder` default values

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:
```diff
@@ -18,6 +18,7 @@ public final class LicenseUtils {
     public enum Product {
         SEARCH_APPLICATION("search application"),
         BEHAVIORAL_ANALYTICS("behavioral analytics"),
+        CONNECTOR("connector"),
```
I think this can be removed now that you're using the `BaseRestHandler`?

kderusso commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorCustomScheduleTests.java:
```diff
@@ -0,0 +1,49 @@
+import static java.util.Collections.emptyList;
+import static org.hamcrest.CoreMatchers.equalTo;
+
+public class ConnectorCustomScheduleTests extends ESTestCase {
```
Nice thorough and explicit test coverage here! 👏

kderusso commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorScheduling.java:
```diff
@@ -0,0 +1,248 @@
+                return this;
+            }
+
+            public Builder setInterval(String interval) {
```
Do we want to do any validation on proper cron syntax here?

jedrazb replied:
Good call, I found [Cron](https://github.com/elastic/elasticsearch/blob/61191b880c7bbcdac6ac8f01a544ca3abe3385a1/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/Cron.java#L200) class in ES which does this already. I will use it and it will handle the validation for us

kderusso reviewed (commented):
Amazing work @jedrazb ! This is great, thanks for iterating!

I especially like the very thorough test coverage. 👏 

I have a few minor comments. I think this is almost ready to approve, pending: 
- Resolving the current String vs. Instance discussion 
- Yaml tests (it can be super simple to start since we only have a PUT command) 

Thank you!

jedrazb committed:
Use Instants to handle datetime, correctly handle nulls, some feedback from Kathleen, fix one unittest

jedrazb committed:
Use Cron class for handling and validatiting Cron interval strings

jedrazb committed:
Add YML test

jedrazb commented:
Updated PR with recent feedback:
- Put the rest endpoints behind a feature flag
- Added YML tests
- Handle intervals as `Cron` and datetime as `Instant` 
- Better handling of null values (writing toXContent)
- Changed how default values are initialized for Connector, everything happens in `Builder`
- All other feedback from @carlosdelest  and @kderusso  should be also addressed

jedrazb committed:
Fix rest-api-spec violation, add temp link for documentation

jedrazb committed:
Merge branch 'main' into jedrazb-connectors-api-connectos-object-and-basic-endpoints

carlosdelest reviewed (approved):
Amazing job! 👏

jedrazb commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/300_connector_put.yml:
```diff
@@ -0,0 +1,47 @@
+
```
Yup that's the plan. Thank you for flagging this

jedrazb commented:
I addressed all feedback from @kderusso - I'm merging this PR (under a feature flag) 🚀  given that Kathleen is OOO for the next few days. It will help me moving forward with the project.
